### ISSUE TITLE
Please help with 'Extected xxx' error types

### ISSUE THREAD

**Issue by thanpolas** [andrewrk/node-multiparty#105]
While trying to debug why our iOS app cannot upload files to our express server using the multiparty package, I stumbled upon a similar issue using Chrome's DHC extension for simulating requests...

The whole of iOS request can be [found here](http://than.pol.as/ZrSh), these are the headers of the iOS request:

```
POST /clip/upload HTTP/1.1
Host: reveal-staging.herokuapp.com
Content-Type: multipart/form-data; boundary=BoundaryJxVkCSjZCIAzwRb
Connection: keep-alive
Accept: application/json
User-Agent: IRClip/1 CFNetwork/711.1.16 Darwin/14.0.0
Content-Length: 479032
Accept-Language: en-us
Accept-Encoding: gzip, deflat
```

And the body is:

```
--BoundaryJxVkCSjZCIAzwRb
Content-Disposition: form-data; camera="iOS"; file="capturedvideo.MOV"
...content...
--BoundaryJxVkCSjZCIAzwRb--
```

This request produces the error:

**Expected alphabetic character, received 0**

When trying to work it out using Chrome's DHC extension I got another type of error:

**Expected CR Received 10**

The request Headers of DHC were:

```
host: 'localhost:3000',
  connection: 'keep-alive',
  'content-length': '154',
  accept: 'application/json',
  origin: 'chrome-extension://aejoelaoggembcahagimdiliamlcdmfm',
  'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.111 Safari/537.36',
  'content-type': 'multipart/form-data; boundary=----WebKitFormBoundaryi3Xz4TKrYpgIdIpf',
  'accept-encoding': 'gzip, deflate',
  'accept-language': 'en,el-GR;q=0.8,el;q=0.6',
```

All tests on the express app, which use *Superagent* pass a-ok, and all browser based uploads work just right.

**Comment by thebigdog** [comment]
For this issue, I got it working by adding some missing fields under the boundary:

Content-Disposition: form-data; camera="iOS"; name="file"; filename="capturedvideo.MOV"
Content-Type: application/octet-stream
Content-Length: 8073196

When I added those it started working properly. The items I changed:

* Under content disposition added name field and changed file to filename
* Added the Content-Type and Content-Length

Hopefully that might help someone in the future.

**Comment by dougwilson** [comment]
@thanpolas since `multipart/form-data` is a binary protocol, I really need you to provide a hex dump of the actual octets, rather than a paste of text. For example, because you pasted it as text, I don't know if what looks like line ending is `0x0d 0x0a` or `0x0a`, which makes a big difference to this protocol.

**Comment by dougwilson** [comment]
@thanpolas I downloaded that `.gz` file you posted and if that's the binary contents of the request, it's definitely invalid `multipart/form-data`--all the parts that should be `CR LF` are all `LF`. My only doubt is that the headers at the top are separated by `LF` instead of the required `CR LF` in HTTP, which makes me think that whatever method you used to capture that was not a raw packet capture at the network level and the contents have been mangled.

**Comment by dougwilson** [comment]
> Expected CR Received 10

This error means that it is starting to read a protocol line ending of `CR LF`, but the data that is being fed into multiparty is just `LF`, which is invalid.

**Comment by dougwilson** [comment]
> Expected alphabetic character, received 0

This error means that the parser expected an alphanumeric character, but received a null `0x00` instead, making it an invalid request. That was most likely caused by the fact that at the end of the `Content-Disposition` line, there was just a `LF` instead of the protocol required `CR LF` and so the header never actually ended, but yet there is a `0x00` character, which is not valid in a header. Specifically your dump shows `Content-Disposition: form-data; camera="iOS"; file="capturedvideo.MOV" [0x0a] [0x00]` when the protocol requires it to be `Content-Disposition: form-data; camera="iOS"; file="capturedvideo.MOV" [0x0d] [0x0a] [0x00]`

**Comment by dougwilson** [comment]
TL;DR your posted errors and the file you provided all point to your issues being something translating all `0x0d 0x0a` sequences to `0x0a` somewhere before multiparty; the iOS client you are using may even be doing this, but suffice to say, this module would never accept those kinds of requests.

**Comment by thanpolas** [comment]
thankyou @dougwilson for the feedback.
